Make `HasDisplayHandle` optional in `WindowHandle`

MarijnS95 opened the pull request:
**Connections**
Followup to https://github.com/gfx-rs/wgpu/pull/8012#issuecomment-3618408387 and https://github.com/rust-windowing/softbuffer/issues/298#issuecomment-3617254907 (CC @cwfitzgerald @dhardy)

**Description**
While true for some window objects like `winit`'s `Window` object, not everyone is or should be required to return their display handle in the same place as the window handle, by means of implementing both `RawDisplayHandle` and `RawWindowHandle` on the same object.

Additionally some APIs, specifically most/all GL(ES) context creation APIs require the `RawDisplayHandle` up-front to create a context that is compatible with that "compositor connection".

For this reason callers should no longer be required to pass a single object that implements both `HasDisplayHandle` and `HasWindowHandle`, allowing the former to become `Option`al in favour of taking the one stored (and kept alive!) in the `Instance`.

`HasDisplayHandle` can still be passed in this position however, because APIs like Vulkan have completely pushed out compositor integration via surfaces to the edge of the API and can technically let one instance or device render into multiple different `RawDisplayHandle` "compositor connections" concurrently.

**Testing**

```console
$ WGPU_BACKEND=gl cargo r -p wgpu-example-02-hello-window
$ WGPU_BACKEND=gl cargo r -p wgpu-examples -- hello_triangle
$ WGPU_BACKEND=gl RUST_LOG=wgpu_hal=debug cargo r -p player # Changed code but don't have a file to play and test it yet
```


**Squash or Rebase?**

Doesn't matter.

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [ ] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [ ] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry.

MarijnS95 commented on examples/standalone/02_hello_window/src/main.rs:
```diff
@@ -32,6 +32,9 @@ impl State {
+        // BREAKING CHANGE: The Into<SurfaceTarget> here will now only
+        // pass the window handle, and fails if the "optional" display
+        // handle wasn't passed above to with_display_handle() too
         let surface = instance.create_surface(window.clone()).unwrap();
```
I'm pretty scared away by pushing this breaking change to users, it's quite "invisible" if one forgets to call `with_display_handle()` on their `InstanceDescriptor` (or replace this call with one that does pass the display handle).

cwfitzgerald replied:
Yeah - do you think changing `InstanceDescriptor::default` (removing the default impl) to `InstanceDescriptor::new` with an optional display handle (or maybe ::with_display_handle()/with_no_display_handle()?) would work? That would make it an explicit breaking change.

MarijnS95 replied:
That seems preferable, should I start deleting the default?

cwfitzgerald replied:
Yeah let's do it

cwfitzgerald replied:
Yeah let's do it

MarijnS95 commented on wgpu/src/api/instance.rs:
```diff
@@ -176,10 +176,19 @@ impl Instance {
+    /// a display handle in [`InstanceDescriptor::display`].
+    ///
+    /// If that cannot be provided, call [`Instance::create_surface_unsafe()`] with a
+    /// [`SurfaceTargetUnsafe::from_display_and_window()`].
```
This is probably a bad fallback, requiring `SurfaceTargetUnsafe` to achieve this unless we want to drive everyone to pass the handle to `InstanceDescriptor`?

Perhaps we should have a non-default `from_window_without_display()` otherwise, if this optionality-by-default isn't preferred.

cwfitzgerald replied:
The three main paths that I expect people to be using (winit, SDL, or raw web handles) can all provide both at the same time, so it seems that providing both by default can't hurt with a separate, safe path for providing the one.

MarijnS95 replied:
I'm not sure how that would look like; restoring the `HasWindowHandle + HasDisplayHandle` bound and trait to be able to satisfy `impl Into<SurfaceTarget>`?

cwfitzgerald replied:
Yeah we have the default Into<SurfaceTarget> be implemented on HasWindowHandle + HasDisplayHandle, then `from_window_without_display` for safe, optional needs.

MarijnS95 replied:
Done and pushed as a fixup. Although now I'm not entirely confident that all the code changes to pass a display through the `InstanceDescriptor` are still necessary, but those codepaths will now simply be able to pass it in two places.

Would love some documentation bikeshedding, I'm loosing context here each time I come back after 3 weeks.

cwfitzgerald commented on examples/standalone/02_hello_window/src/main.rs:
```diff
@@ -32,6 +32,9 @@ impl State {
 
+        // BREAKING CHANGE: The Into<SurfaceTarget> here will now only
+        // pass the window handle, and fails if the "optional" display
+        // handle wasn't passed above to with_display_handle() too
```
This comment definitely doesn't belong here, though it was useful in calling attention to the exact change that was being made here.

cwfitzgerald commented on wgpu-core/src/instance.rs:
```diff
@@ -932,6 +939,8 @@ pub enum CreateSurfaceError {
     FailedToCreateSurfaceForAnyBackend(HashMap<Backend, hal::InstanceError>),
     #[error("The display handle used to create this Instance does not match the one used to create a surface on it")]
     MismatchingDisplayHandle,
+    #[error("Neither the `Instance` nor `create_surface()` parameters received a `DisplayHandle`")]
```
This is an error that people are going to hit and be extremely confused by, so I'm going to expand this error quite a bit:


```suggestion
    #[error(
        "No `DisplayHandle` is available to create this surface with. When creating a surface with `create_surface()` \
        you must specify a display handle in InstanceDescriptor::display. \
        Rarely, if you need to create surfaces from different `DisplayHandle`s (ex. different wayland or x11 connections), \
        you must use create_surface_unsafe()."
    )]
```

cwfitzgerald commented on wgpu-hal/src/gles/wgl.rs:
```diff
@@ -563,12 +563,12 @@ impl crate::Instance for Instance {
-    #[cfg_attr(target_os = "macos", allow(unused, unused_mut, unreachable_code))]
     unsafe fn create_surface(
         &self,
         _display_handle: RawDisplayHandle,
```
Nit - _display_handle should have the underscore removed, it's no longer unused.

MarijnS95 replied:
Yeah well, there wasn't a lot of intent to have these forced assertions, just to prevent some "weirdness". This is all gone in the `glutin` variant though.

cwfitzgerald commented on wgpu/src/api/instance.rs:
```diff
@@ -191,6 +200,8 @@ impl Instance {
             SurfaceTarget::Window(window) => unsafe {
                 let surface = self.create_surface_unsafe(
+                    // TODO: Could also call from_display_and_window depending on the argument;
+                    // currently this is a hidden breaking change.
```
Wouldn't this require specialization?

MarijnS95 replied:
Since the removal of `HasDisplayHandle` from `WindowHandle` (used in `SurfaceTarget::Window`), yes this won't be possible.

cwfitzgerald commented on wgpu/src/api/surface.rs:
```diff
@@ -309,16 +312,19 @@ pub enum SurfaceTargetUnsafe {
     /// will not be supported by any adapters.
     ///
+    /// If the `raw_display_handle` is not [`None`] here and was not [`None`] in
+    /// [`wgt::InstanceDescriptor::display`], their values _must_ be identical.
```
```suggestion
    /// [`wgpu::InstanceDescriptor::display`], their values _must_ be identical.
```

MarijnS95 replied:
`wgpu` isn't in scope because that's the name of the current crate. It must be referenced via `crate::` then because it's reexported from `wgt`.

cwfitzgerald commented on CHANGELOG.md:
```diff
@@ -63,6 +63,10 @@ Bottom level categories:
 
 - Expanded documentation of `QuerySet`, `QueryType`, and `resolve_query_set()` describing how to use queries. By @kpreid in [#8776](https://github.com/gfx-rs/wgpu/pull/8776).
 
+### Hacks (for lack of a better description)
```
Not really a hack, just an API change.

This can live in Changes > General - we'll need to lift it to major changes with a bunch of exposition - but I'm waiting to do that until all the various surface related changes land first.

cwfitzgerald commented on deno_webgpu/byow.rs:
```diff
@@ -137,7 +137,8 @@ impl UnsafeWindowSurface {
     let id = unsafe {
       instance
-        .instance_create_surface(display_handle, win_handle, None)
+        // TODO: May need to be None in some cases
```
Hum?

MarijnS95 replied:
In case the underlying `instance` might want to use the already-stored `display` handle; but then this `byow` API around `fn raw_window()` and `UnsafeWindowSurfaceOptions::display_handle` just needs to change.

I just realized turning that field into an `Option` would represent the change made in this API; would that align?

cwfitzgerald replied:
Yeah that seems fine.

cwfitzgerald reviewed (changes requested):
Broadly good, mostly small docs/error issues

MarijnS95 commented on wgpu-hal/src/gles/egl.rs:
```diff
@@ -888,7 +888,6 @@ impl crate::Instance for Instance {
     }
 
-    #[cfg_attr(target_os = "macos", allow(unused, unused_mut, unreachable_code))]
     unsafe fn create_surface(
```
Wasn't sure of this specific copy-paste bug, there's no `macos` conditional code in this function so no need to conditionally allow lints here.

`expect()` would have helped.

MarijnS95 commented on wgpu/src/api/instance.rs:
```diff
@@ -227,13 +235,12 @@ impl Instance {
                 let obj = core::ptr::NonNull::from(value).cast();
                 let raw_window_handle =
                     raw_window_handle::WebOffscreenCanvasWindowHandle::new(obj).into();
-                let raw_display_handle = raw_window_handle::WebDisplayHandle::new().into();
 
                 // Note that we need to call this while we still have `value` around.
                 // This is safe without storing canvas to `handle_origin` since the surface will create a copy internally.
                 unsafe {
                     self.create_surface_unsafe(SurfaceTargetUnsafe::RawHandle {
-                        raw_display_handle,
+                        raw_display_handle: None,
                         raw_window_handle,
```
It's a bit annoying that now because of this change, such an opaque empty `WebDisplayHandle` needs to be passed to `wgpu` otherwise also the web tests will fail; even if the object is not needed.

On the other hand it might be useful to distinguish the current platform?

MarijnS95 replied:
To fix [those resulting CI failures](https://github.com/gfx-rs/wgpu/actions/runs/20699726842/job/59420158890?pr=8782) I wanted to set a `WebDisplayHandle` in `tests/src/init.rs` instead (because there's a conditional that uses a canvas as surface only on `wasm32` with `emscripten` or `webgl`). Unfortunately `WebDisplayHandle` doesn't implement `HasDisplayHandle`, and while `DisplayHandle` does (and can easily be constructed with `DisplayHandle::web()`) it doesn't satisfy the `Send`+`Sync` requirement we have for `WgpuHasDisplayHandle`.

Easy to work around, but annoying (to the point where we might want to type up edge cases or not? It all gets very confusing and messy)

cwfitzgerald replied:
> (to the point where we might want to type up edge cases or not? It all gets very confusing and messy)

Yeah I'm even have to think quite a bit to figure out what's going on. Definitely worth calling out in documentation explicitly.

MarijnS95 committed:
Make `HasDisplayHandle` optional in `WindowHandle`

While true for some window objects like `winit`'s `Window` object,
not everyone is or should be required to return their display handle
in the same place as the window handle, by means of implementing both
`RawDisplayHandle` and `RawWindowHandle` on the same object.

Additionally some APIs, specifically most/all GL(ES) context creation
APIs require the `RawDisplayHandle` up-front to create a context that is
compatible with that "compositor connection".

For this reason callers should no longer be required to pass a single
object that implements both `HasDisplayHandle` and `HasWindowHandle`,
allowing the former to become `Option`al in favour of taking the one
stored (and kept alive!) in the `Instance`.

`HasDisplayHandle` can still be passed in this position however, because
APIs like Vulkan have completely pushed out compositor integration via
surfaces to the edge of the API and can technically let one instance
or device render into multiple different `RawDisplayHandle` "compositor
connections" concurrently.

MarijnS95 commented on wgpu-types/src/instance.rs:
Just noticing that this should have been:


```suggestion
            display: self.display,
```

A mistake from the previous PR 😬

MarijnS95 committed:
WIP: Differentiate `InstanceDescriptor` constructors with and without display handle

cwfitzgerald commented:
Alright, once the interface is touched up, I think this is good to land :)

cwfitzgerald commented on wgpu/src/api/instance.rs:
```diff
@@ -47,7 +47,8 @@ impl Default for Instance {
     /// this method will panic, see [`Instance::enabled_backend_features()`].
     fn default() -> Self {
-        Self::new(InstanceDescriptor::default())
+        // TODO: Differentiate constructors here too?
```
I had no idea this implementation actually existed. I think we should just remove it.

MarijnS95 replied:
If I remove it, at least `9` compiler errors across various samples that call `wgpu::Instance::default()`... :sweat_smile: 

These refactors of the core API surface all feel a bit out of my league, I'd rather move on with the `glutin` rewrite :wink:

cwfitzgerald replied:
Oop, didn't realize, yeah lets leave as is.

cwfitzgerald commented on CHANGELOG.md:
```diff
@@ -122,6 +122,8 @@ depth_stencil: Some(wgpu::DepthStencilState::stencil(
 - BREAKING: Add `V6_8` variant to `DxcShaderModel` and `naga::back::hlsl::ShaderModel`. By @inner-daemons in [#8882](https://github.com/gfx-rs/wgpu/pull/8882) and @ErichDonGubler in [#9083](https://github.com/gfx-rs/wgpu/pull/9083).
 - BREAKING: Add `V6_9` variant to `DxcShaderModel` and `naga::back::hlsl::ShaderModel`. By @ErichDonGubler in [#????](https://github.com/gfx-rs/wgpu/pull/????).
+- `DisplayHandle` is now optional in surface creation if it was passed to `InstanceDescriptor::display`. By @MarijnS95 in [#8782](https://github.com/gfx-rs/wgpu/pull/8782)
+- TODO: Describe with_env and from_env_or_default refactors if accepted
```
This needs updating with the current state of changes and moved to the unreleased section.

MarijnS95 replied:
Nothing to describe anymore I think given that we restored the original trait that pulls out a `Display` handle from a "`Window`" when creating a `Surface`, unless this document is the right place to be really specific about all the (mostly) new/additional APIs?

It should already be under `## Unreleased`, right?

cwfitzgerald replied:
Yeah seems fine now, for some reason our changelog job was mad previously, seems fine.

> unless this document is the right place to be really specific about all the (mostly) new/additional APIs?

I will probably add some more description of APIs at release time, lets not block this PR.

MarijnS95 replied:
Thanks, yeah I'd rather let you take care of it (and give the API an "external" once-over to write it down, outside of my tunnel vision at this point).

cwfitzgerald reviewed (approved):
Alright, I think the docs are fine. Lets get the changelog updated and get this landed, we can iterate later if needd.

cwfitzgerald commented:
That's everything! Lets land this!
